AlleleCallEvaluator - Build an interactive report for allele calling results evaluation
=======================================================================================

The AlleleCallEvaluator module allows users to generate an interactive HTML report to evaluate
allele calling results generated by the :doc:`AlleleCall </user/modules/AlleleCall>` module. The
report provides summary statistics to evaluate results per sample and per locus (with the possibility
to provide a TSV file with loci annotations to include on a table). The report includes components
to display a heatmap representing the Loci Presence-Absence matrix, a heatmap representing the
Distance matrix based on allelic differences and a Neighbor-Joining (NJ) tree based on the core-genome
MLST alignment.

Basic Usage
:::::::::::

::

	chewBBACA.py AlleleCallEvaluator -i /path/to/AlleleCallResultsFolder -g /path/to/SchemaDirectory -o /path/to/OutputFolderName --cpu 4

Parameters
::::::::::

::

    -i, --input-files           (Required) Path to the directory that contains the allele calling results (default: None).

    -g, --schema-directory      (Required) Path to the schema directory (default: None).

    -o, --output-directory      (Optional) Path to the output directory where the report HTML files
                                will be generated (default: None).

    -a, --annotations           (Optional) Path to the TSV file created by the UniprotFinder module (
                                default: None).

    --cpu, --cpu-cores          (Optional) Number of CPU cores/threads that will be used to run the
                                process (will be redefined to a lower value if it is equal to or
                                exceeds the totalnumber of available CPU cores/threads) (default: 1).

    --light                     (Optional) Do not compute presence-absence matrix, distance matrix
                                and the NJ cgMLST tree (default: False).

    --no-pa                     (Optional) Do not compute the presence-absence matrix (default: False).

    --no-dm                     (Optional) Do not compute the distance matrix (default: False).

    --no-pa                     (Optional) Do not compute the NJ cgMLST tree (default: False).

    --no-pa                     (Optional) Compute the cgMLST alignment, even if `--no-tree` is True (default: False).

Outputs
:::::::

::

   OutputFolderName
   ├── allelic_differences_symmetric.tsv
   ├── cgMLST_alignment.fasta
   ├── main_report.html
   ├── masked.tsv
   ├── Presence_Absence.tsv
   └── report_bundle.js

- A TSV file, ``allelic_differences_symmetric.tsv``, that contains the symmetric distance matrix based on allelic differences.

- A FASTA file, ``cgMLST_alignment.fasta``, that contains the core-genome MLST alignment.

- A HTML report, ``main_report.html``, that contains the following components:

  - A table with the total number of samples, total number of loci, total number of coding sequences (CDSs) extracted from the samples, total number of CDSs classified and totals per classification type.
  - A tab panel with stacked bar charts for the classification counts per sample and per locus.
  - A tab panel with more detailed sample and locus statistics.
  - If a TSV file with annotations is provided to the ``--annotations`` parameter, the report
    will also include a table with the provided annotations. Otherwise, it will display a warning informing that
    no annotations were provided.
  - A Heatmap chart representing the loci presence-absence matrix for all samples in the dataset.
  - A Heatmap chart representing the allelic distance matrix for all samples in the dataset.
  - A tree drawn with Phylocanvas based on the Neighbor-Joining (NJ) tree computed by FastTree.

- A TSV file, ``masked.tsv``, that contains the masked allelic profiles (results from masking the allelic profiles in the results_alleles.tsv file generated by the AlleleCall module).

- A TSV file, ``Presence_Absence.tsv``, that contains the loci presence-absence matrix.

- A JavaScript bundle file, ``report_bundle.js``.

.. warning::
  The JS bundle is necessary to visualize the HTML reports. Do not delete this file. You should
  not move or delete this file. If you want to share the report, simply
  compress the output folder and share the compressed archive. The receiver can simply uncompress
  the archive and open the HTML file in a browser to visualize the report.

Report Components
-----------------

The first component gives a small introduction that details the type of information contained in
each component of the schema report.

.. image:: /_static/images/allelecall_report_description.png
   :width: 1400px
   :align: center

Results Summary Data
....................

The second component is a table with summary statistics about the allele calling results, such as:

  - **Total Samples**: Total number of samples in the dataset.
  - **Total Loci**: Total number of loci used to perform allele calling.
  - **Total CDSs**: Total number of coding sequences identified in all the samples.
  - **Total CDSs Classified**: Total number of CDSs that were classified.
  - **EXC**: Total number of CDSs classified as EXC.
  - **INF**: Total number of CDSs classified as INF.
  - **PLOT3**: Total number of CDSs classified as PLOT3.
  - **PLOT5**: Total number of CDSs classified as PLOT5.
  - **LOTSC**: Total number of CDSs classified as LOTSC.
  - **NIPH**: Total number of NIPH classifications (each NIPH classification includes at least two CDSs).
  - **NIPHEM**: Total number of NIPHEM classifications (each NIPHEM classification includes at least two CDSs).
  - **ALM**: Total number of CDSs classified as ALM.
  - **ASM**: Total number of CDSs classified as ASM.
  - **PAMA**: Total number of PAMA classifications (each PAMA classification includes at least two CDSs).

.. image:: /_static/images/allelecall_report_summary.png
   :width: 1400px
   :align: center

Please visit the section about the :doc:`AlleleCall </user/modules/AlleleCall>` module if you want to know
more about each classification type.

Classification Counts
.....................

The third component contains 2 panels with stacked bar charts displaying the classification counts
per sample and per locus.

- Panel A, ``Counts Per Sample``, displays the stacked bar charts for the sample classification counts.

.. image:: /_static/images/allelecall_report_sample_counts.png
   :width: 1400px
   :align: center

- Panel B, ``Counts Per Locus``, displays the stacked bar charts for the loci classification counts.

.. image:: /_static/images/allelecall_report_loci_counts.png
   :width: 1400px
   :align: center

Detailed Statistics
...................

The fourth component contains 2 panels with tables with detailed statistics about the results per sample and per locus.
The dropdown menu below the tables allows the selection of a single column to generate a histogram for the values in
the selected column.

.. image:: /_static/images/allelecall_report_detailed_stats.png
   :width: 1400px
   :align: center


Loci annotations
................

If a TSV file with loci annotations is provided, the fifth component of the schema report is a table
with the list of annotations. Otherwise, it will display a warning informing that no annotations
were provided.

.. image:: /_static/images/allelecall_report_annotations.png
   :width: 1400px
   :align: center

If a column name includes ``URL``, the AlleleCallEvaluator module assumes that the values in that column
are URLs and creates links to the web pages.

.. important::
  The first column in the TSV file with annotations must be named ``Locus`` and contain the identifiers
  of the loci (the basename of the locus FASTA file without the ``.fasta`` extension).

You can use the :doc:`UniprotFinder </user/modules/UniprotFinder>` module to annotate the loci in a schema
created with chewBBACA. If you want to annotate an external schema, you can adapt it with the
:doc:`PrepExternalSchema </user/modules/PrepExternalSchema>` module followed by annotation with the
:doc:`UniprotFinder </user/modules/UniprotFinder>` module.

Loci Presence-Absence
.....................

The sixth component displays a Heatmap representing the loci presence-absence matrix for all samples in the
dataset. Blue cells (z=1) correspond to loci presence and grey cells (z=0) to loci absence. The "Select Sample"
dropdown menu enables the selection of a single sample to display its heatmap on top of the main heatmap. The
"Select Locus" dropdown menu enables the selection of a single locus to display its heatmap on the right of
the main heatmap.

.. image:: /_static/images/allelecall_report_pa_heatmap.png
   :width: 1400px
   :align: center


Allelic Distances
.................

The seventh component displays a Heatmap representing the allelic distance matrix for all samples in the dataset.
The allelic distances are computed based on the cgMLST profiles (loci that are not present in all samples are not
included in the computations). The "Select Sample" dropdown menu enables the selection of a single sample to display
its heatmap on top of the main heatmap. The menu after the heatmap enables the selection of a single sample and of
a distance threshold to display a table with the list of samples at a distance equal or smaller than the specified
distance value.

.. image:: /_static/images/allelecall_report_dm_heatmap.png
   :width: 1400px
   :align: center


Allelic Distances
.................

The last component displays a tree drawn with Phylocanvas based on the Neighbor-Joining (NJ) tree computed by FastTree.
The tree is computed based on the MSA for the set of loci that constitute the core-genome.

.. image:: /_static/images/allelecall_report_cgMLST_tree.png
   :width: 1400px
   :align: center
